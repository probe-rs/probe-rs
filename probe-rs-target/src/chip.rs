use super::memory::MemoryRegion;
use crate::{serialize::hex_option, CoreType};
use serde::{Deserialize, Serialize};

/// Represents a DAP scan chain element.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct ScanChainElement {
    /// Unique name of the DAP
    pub name: Option<String>,
    /// Specifies the IR length of the DAP (default value: 4).
    pub ir_len: Option<u8>,
}

/// A finite list of all possible binary formats a target might support.
#[derive(Debug, Default, Serialize, Deserialize, PartialEq, Eq, Clone)]
#[serde(rename_all = "lowercase")]
pub enum BinaryFormat {
    /// Program sections are bit-for-bit copied to flash.
    #[default]
    Raw,
    /// Program sections are copied to flash, with the relevant headers and metadata for the [ESP-IDF bootloader](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/app_image_format.html#app-image-structures).
    Idf,
}

/// Configuration for JTAG probes.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Jtag {
    /// Describes the scan chain
    ///
    /// ref: `<https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/sdf_pg.html#sdf_element_scanchain>`
    #[serde(default)]
    pub scan_chain: Option<Vec<ScanChainElement>>,
}

/// A single chip variant.
///
/// This describes an exact chip variant, including the cores, flash and memory size. For example,
/// the `nRF52832` chip has two variants, `nRF52832_xxAA` and `nRF52832_xxBB`. For this case,
/// the struct will correspond to one of the variants, e.g. `nRF52832_xxAA`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Chip {
    /// This is the name of the chip in base form.
    /// E.g. `nRF52832`.
    pub name: String,
    /// The `PART` register of the chip.
    /// This value can be determined via the `cli info` command.
    pub part: Option<u16>,
    /// An URL to the SVD file for this chip.
    pub svd: Option<String>,
    /// The cores available on the chip.
    #[serde(default)]
    pub cores: Vec<Core>,
    /// The memory regions available on the chip.
    pub memory_map: Vec<MemoryRegion>,
    /// Names of all flash algorithms available for this chip.
    ///
    /// This can be used to look up the flash algorithm in the
    /// [`ChipFamily::flash_algorithms`] field.
    ///
    /// [`ChipFamily::flash_algorithms`]: crate::ChipFamily::flash_algorithms
    #[serde(default)]
    pub flash_algorithms: Vec<String>,
    /// Specific memory ranges to search for a dynamic RTT header for code
    /// running on this chip.
    ///
    /// This need not be specified for most chips because the default is
    /// to search all RAM regions specified in `memory_map`. However,
    /// that behavior isn't appropriate for some chips, such as those which
    /// have a very large amount of RAM that would be time-consuming to
    /// scan exhaustively.
    ///
    /// If specified then this is a list of zero or more address ranges to
    /// scan. Each address range must be enclosed in exactly one RAM region
    /// from `memory_map`. An empty list disables automatic scanning
    /// altogether, in which case RTT will be enabled only when using an
    /// executable image that includes the `_SEGGER_RTT` symbol pointing
    /// to the exact address of the RTT header.
    pub rtt_scan_ranges: Option<Vec<std::ops::Range<u64>>>,
    /// JTAG-specific options
    #[serde(default)]
    pub jtag: Option<Jtag>,
    /// The default binary format for this chip
    pub default_binary_format: Option<BinaryFormat>,
}

impl Chip {
    /// Create a generic chip with the given name, a single core,
    /// and no flash algorithm or memory map. Used to create
    /// generic targets.
    pub fn generic_arm(name: &str, core_type: CoreType) -> Self {
        Chip {
            name: name.to_string(),
            part: None,
            svd: None,
            cores: vec![Core {
                name: "main".to_string(),
                core_type,
                core_access_options: CoreAccessOptions::Arm(ArmCoreAccessOptions::default()),
            }],
            memory_map: vec![],
            flash_algorithms: vec![],
            rtt_scan_ranges: None,
            jtag: None,
            default_binary_format: Some(BinaryFormat::Raw),
        }
    }
}

/// An individual core inside a chip
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Core {
    /// The core name.
    pub name: String,

    /// The core type.
    /// E.g. `M0` or `M4`.
    #[serde(rename = "type")]
    pub core_type: CoreType,

    /// The AP number to access the core
    pub core_access_options: CoreAccessOptions,
}

/// The data required to access a core
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum CoreAccessOptions {
    /// ARM specific options
    Arm(ArmCoreAccessOptions),
    /// RISC-V specific options
    Riscv(RiscvCoreAccessOptions),
    /// Xtensa specific options
    Xtensa(XtensaCoreAccessOptions),
}

/// The data required to access an ARM core
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ArmCoreAccessOptions {
    /// The access port number to access the core
    pub ap: u8,
    /// The port select number to access the core
    pub psel: u32,
    /// The base address of the debug registers for the core.
    /// Required for Cortex-A, optional for Cortex-M
    #[serde(serialize_with = "hex_option")]
    pub debug_base: Option<u64>,
    /// The base address of the cross trigger interface (CTI) for the core.
    /// Required in ARMv8-A
    #[serde(serialize_with = "hex_option")]
    pub cti_base: Option<u64>,
}

/// The data required to access a Risc-V core
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RiscvCoreAccessOptions {
    /// The hart id
    pub hart_id: Option<u32>,

    /// The JTAG TAP index of the core's debug module
    pub jtag_tap: Option<usize>,
}

/// The data required to access an Xtensa core
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct XtensaCoreAccessOptions {
    /// The JTAG TAP index of the core's debug module
    pub jtag_tap: Option<usize>,
}
