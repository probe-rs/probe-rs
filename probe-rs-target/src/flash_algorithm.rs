use super::flash_properties::FlashProperties;
use crate::serialize::{hex_option, hex_u_int};
use base64::{engine::general_purpose as base64_engine, Engine as _};
use serde::{Deserialize, Serialize};

/// Data encoding used by the flash algorithm.
#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum TransferEncoding {
    /// Raw binary encoding. Probe-rs will not apply any transformation to the flash data.
    #[default]
    Raw,

    /// Flash data is compressed using the `miniz_oxide` crate.
    ///
    /// Compressed images are written in page sized chunks, each chunk written to the image's start
    /// address. The length of the compressed image is stored in the first 4 bytes of the first
    /// chunk of the image.
    Miniz,
}

/// The raw flash algorithm is the description of a flash algorithm,
/// and is usually read from a target description file.
///
/// Before it can be used for flashing, it has to be assembled for
/// a specific chip, by determining the RAM addresses which are used when flashing.
/// This process is done in the main `probe-rs` library.
#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct RawFlashAlgorithm {
    /// The name of the flash algorithm.
    pub name: String,
    /// The description of the algorithm.
    pub description: String,
    /// Whether this flash algorithm is the default one or not.
    #[serde(default)]
    pub default: bool,
    /// List of 32-bit words containing the code for the algo. If `load_address` is not specified, the code must be position independent (PIC).
    #[serde(deserialize_with = "deserialize")]
    #[serde(serialize_with = "serialize")]
    pub instructions: Vec<u8>,
    /// Address to load algo into RAM. Optional.
    #[serde(serialize_with = "hex_option")]
    pub load_address: Option<u64>,
    /// Address to load data into RAM. Optional.
    #[serde(serialize_with = "hex_option")]
    pub data_load_address: Option<u64>,
    /// Address of the `Init()` entry point. Optional.
    #[serde(serialize_with = "hex_option")]
    pub pc_init: Option<u64>,
    /// Address of the `UnInit()` entry point. Optional.
    #[serde(serialize_with = "hex_option")]
    pub pc_uninit: Option<u64>,
    /// Address of the `ProgramPage()` entry point.
    #[serde(serialize_with = "hex_u_int")]
    pub pc_program_page: u64,
    /// Address of the `EraseSector()` entry point.
    #[serde(serialize_with = "hex_u_int")]
    pub pc_erase_sector: u64,
    /// Address of the `EraseAll()` entry point. Optional.
    #[serde(serialize_with = "hex_option")]
    pub pc_erase_all: Option<u64>,
    /// The offset from the start of RAM to the data section.
    #[serde(serialize_with = "hex_u_int")]
    pub data_section_offset: u64,
    /// Location of the RTT control block in RAM.
    ///
    /// If this is set, the flash algorithm supports RTT output
    /// and debug messages will be read over RTT.
    #[serde(serialize_with = "hex_option")]
    pub rtt_location: Option<u64>,
    /// The properties of the flash on the device.
    pub flash_properties: FlashProperties,
    /// List of cores that can use this algorithm
    #[serde(default)]
    pub cores: Vec<String>,
    /// The flash algorithm's stack size, in bytes.
    ///
    /// If not set, probe-rs selects a default value.
    /// Increase this value if you're concerned about stack
    /// overruns during flashing.
    pub stack_size: Option<u32>,

    /// The encoding format accepted by the flash algorithm.
    #[serde(default)]
    pub transfer_encoding: Option<TransferEncoding>,
}

pub fn serialize<S>(bytes: &[u8], serializer: S) -> Result<S::Ok, S::Error>
where
    S: serde::Serializer,
{
    serializer.serialize_str(base64_engine::STANDARD.encode(bytes).as_str())
}

pub fn deserialize<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    struct Base64Visitor;

    impl<'de> serde::de::Visitor<'de> for Base64Visitor {
        type Value = Vec<u8>;

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            write!(formatter, "base64 ASCII text")
        }

        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            base64_engine::STANDARD
                .decode(v)
                .map_err(serde::de::Error::custom)
        }
    }

    deserializer.deserialize_str(Base64Visitor)
}
