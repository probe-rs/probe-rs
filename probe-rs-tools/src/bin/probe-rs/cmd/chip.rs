use bytesize::ByteSize;
use probe_rs::config::MemoryRegion;

#[derive(clap::Parser)]
pub struct Cmd {
    #[clap(subcommand)]
    subcommand: Subcommand,
}

#[derive(clap::Subcommand)]
/// Inspect internal registry of supported chips
enum Subcommand {
    /// Lists all the available families and their chips with their full.
    #[clap(name = "list")]
    List,
    /// Shows chip properties of a specific chip
    #[clap(name = "info")]
    Info {
        /// The name of the chip to display.
        name: String,
    },
}

impl Cmd {
    pub fn run(self) -> anyhow::Result<()> {
        let output = std::io::stdout().lock();

        match self.subcommand {
            Subcommand::List => print_families(output),
            Subcommand::Info { name } => print_chip_info(output, &name),
        }
    }
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_families(mut output: impl std::io::Write) -> anyhow::Result<()> {
    writeln!(output, "Available chips:")?;
    for family in probe_rs::config::families() {
        writeln!(output, "{}", &family.name)?;
        writeln!(output, "    Variants:")?;
        for variant in family.variants() {
            writeln!(output, "        {}", variant.name)?;
        }
    }
    Ok(())
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_chip_info(mut output: impl std::io::Write, name: &str) -> anyhow::Result<()> {
    writeln!(output, "{}", name)?;
    let target = probe_rs::config::get_target_by_name(name)?;
    writeln!(output, "Cores ({}):", target.cores.len())?;
    for core in target.cores {
        writeln!(
            output,
            "    - {} ({:?})",
            core.name.to_ascii_lowercase(),
            core.core_type
        )?;
    }

    fn get_range_len(range: &std::ops::Range<u64>) -> u64 {
        range.end - range.start
    }

    for memory in target.memory_map {
        let range = memory.address_range();
        let size = ByteSize(get_range_len(&range)).to_string_as(true);
        let kind = match memory {
            MemoryRegion::Ram(_) => "RAM",
            MemoryRegion::Generic(_) => "Generic",
            MemoryRegion::Nvm(_) => "NVM",
        };
        writeln!(output, "{kind}: {range:#010x?} ({size})")?
    }
    Ok(())
}

#[test]
fn single_chip_output() {
    let mut buff = Vec::new();
    print_chip_info(&mut buff, "nrf52840_xxaa").unwrap();

    // output should be valid utf8
    let output = String::from_utf8(buff).unwrap();

    insta::assert_snapshot!(output);
}

#[test]
fn multiple_chip_output() {
    let mut buff = Vec::new();
    let error = print_chip_info(&mut buff, "nrf52").unwrap_err();

    insta::assert_snapshot!(error.to_string());
}
