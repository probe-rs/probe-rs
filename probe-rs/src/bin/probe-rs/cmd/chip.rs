use byte_unit::Byte;

#[derive(clap::Parser)]
pub struct Cmd {
    #[clap(subcommand)]
    subcommand: Subcommand,
}

#[derive(clap::Subcommand)]
/// Inspect internal registry of supported chips
enum Subcommand {
    /// Lists all the available families and their chips with their full.
    #[clap(name = "list")]
    List,
    /// Shows chip properties of a specific chip
    #[clap(name = "info")]
    Info {
        /// The name of the chip to display.
        name: String,
    },
}

impl Cmd {
    pub fn run(self) -> anyhow::Result<()> {
        let output = std::io::stdout().lock();

        match self.subcommand {
            Subcommand::List => print_families(output),
            Subcommand::Info { name } => print_chip_info(output, &name),
        }
    }
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_families(mut output: impl std::io::Write) -> anyhow::Result<()> {
    writeln!(output, "Available chips:")?;
    for family in probe_rs::config::families() {
        writeln!(output, "{}", &family.name)?;
        writeln!(output, "    Variants:")?;
        for variant in family.variants() {
            writeln!(output, "        {}", variant.name)?;
        }
    }
    Ok(())
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_chip_info(mut output: impl std::io::Write, name: &str) -> anyhow::Result<()> {
    writeln!(output, "{}", name)?;
    let target = probe_rs::config::get_target_by_name(name)?;
    writeln!(output, "Cores ({}):", target.cores.len())?;
    for core in target.cores {
        writeln!(
            output,
            "    - {} ({:?})",
            core.name.to_ascii_lowercase(),
            core.core_type
        )?;
    }

    fn get_range_len(range: &std::ops::Range<u64>) -> u64 {
        range.end - range.start
    }

    for memory in target.memory_map {
        match memory {
            probe_rs::config::MemoryRegion::Ram(region) => writeln!(
                output,
                "RAM: {:#010x?} ({:.2})",
                &region.range,
                Byte::from_u64(get_range_len(&region.range))
                    .get_appropriate_unit(byte_unit::UnitType::Binary)
            )?,
            probe_rs::config::MemoryRegion::Generic(region) => writeln!(
                output,
                "Generic: {:#010x?} ({:.2})",
                &region.range,
                Byte::from_u64(get_range_len(&region.range))
                    .get_appropriate_unit(byte_unit::UnitType::Binary)
            )?,
            probe_rs::config::MemoryRegion::Nvm(region) => writeln!(
                output,
                "NVM: {:#010x?} ({:.2})",
                &region.range,
                Byte::from_u64(get_range_len(&region.range))
                    .get_appropriate_unit(byte_unit::UnitType::Binary)
            )?,
        };
    }
    Ok(())
}

#[test]
fn single_chip_output() {
    let mut buff = Vec::new();
    print_chip_info(&mut buff, "nrf52840_xxaa").unwrap();

    // output should be valid utf8
    let output = String::from_utf8(buff).unwrap();

    insta::assert_snapshot!(output);
}

#[test]
fn multiple_chip_output() {
    let mut buff = Vec::new();
    let error = print_chip_info(&mut buff, "nrf52").unwrap_err();

    insta::assert_snapshot!(error.to_string());
}
