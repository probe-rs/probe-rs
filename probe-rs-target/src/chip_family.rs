use std::{
    collections::HashMap,
    fs::{read, read_to_string},
    path::{Path, PathBuf},
};

use crate::CoreAccessOptions;

use super::chip::Chip;
use super::flash_algorithm::RawFlashAlgorithm;
use base64::Engine;
use jep106::JEP106Code;

use serde::{Deserialize, Serialize};
use serde_yaml::{Mapping, Value};

/// Source of a target description.
///
/// This is used for diagnostics, when
/// an error related to a target description occurs.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum TargetDescriptionSource {
    /// The target description is a generic target description,
    /// which just describes a core type (e.g. M4), without any
    /// flash algorithm or memory description.
    Generic,
    /// The target description is a built-in target description,
    /// which was included into probe-rs at compile time.
    BuiltIn,
    /// The target description was from an external source
    /// during runtime.
    External,
}

/// Type of a supported core.
#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CoreType {
    /// ARMv6-M: Cortex M0, M0+, M1
    Armv6m,
    /// ARMv7-A: Cortex A7, A9, A15
    Armv7a,
    /// ARMv7-M: Cortex M3
    Armv7m,
    /// ARMv7e-M: Cortex M4, M7
    Armv7em,
    /// ARMv7-A: Cortex A35, A55, A72
    Armv8a,
    /// ARMv8-M: Cortex M23, M33
    Armv8m,
    /// RISC-V
    Riscv,
    /// Xtensa - TODO: may need to split into NX, LX6 and LX7
    Xtensa,
}

impl CoreType {
    /// Returns true if the core type is an ARM Cortex-M
    pub fn is_cortex_m(&self) -> bool {
        matches!(
            self,
            CoreType::Armv6m | CoreType::Armv7em | CoreType::Armv7m | CoreType::Armv8m
        )
    }
}

/// The architecture family of a specific [`CoreType`].
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Architecture {
    /// An ARM core of one of the specific types [`CoreType::Armv6m`], [`CoreType::Armv7m`], [`CoreType::Armv7em`] or [`CoreType::Armv8m`]
    Arm,
    /// A RISC-V core.
    Riscv,
    /// An Xtensa core.
    Xtensa,
}

impl CoreType {
    /// Returns the parent architecture family of this core type.
    pub fn architecture(&self) -> Architecture {
        match self {
            CoreType::Riscv => Architecture::Riscv,
            CoreType::Xtensa => Architecture::Xtensa,
            _ => Architecture::Arm,
        }
    }
}

/// Instruction set used by a core
#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum InstructionSet {
    /// ARM Thumb 2 instruction set
    Thumb2,
    /// ARM A32 (often just called ARM) instruction set
    A32,
    /// ARM A64 (aarch64) instruction set
    A64,
    /// RISC-V 32-bit uncompressed instruction sets (RV32) - covers all ISA variants that use 32-bit instructions.
    RV32,
    /// RISC-V 32-bit compressed instruction sets (RV32C) - covers all ISA variants that allow compressed 16-bit instructions.
    RV32C,
    /// Xtensa instruction set
    Xtensa,
}

impl InstructionSet {
    /// Get the instruction set from a rustc target triple.
    pub fn from_target_triple(triple: &str) -> Option<Self> {
        match triple.split('-').next()? {
            "thumbv6m" | "thumbv7em" | "thumbv7m" | "thumbv8m" => Some(InstructionSet::Thumb2),
            "arm" => Some(InstructionSet::A32),
            "aarch64" => Some(InstructionSet::A64),
            "xtensa" => Some(InstructionSet::Xtensa),
            other => {
                if let Some(features) = other.strip_prefix("riscv32") {
                    if features.contains('c') {
                        Some(InstructionSet::RV32C)
                    } else {
                        Some(InstructionSet::RV32)
                    }
                } else {
                    None
                }
            }
        }
    }

    /// Get the minimum instruction size in bytes.
    pub fn get_minimum_instruction_size(&self) -> u8 {
        match self {
            InstructionSet::Thumb2 => {
                // Thumb2 uses a variable size (2 or 4) instruction set. For our purposes, we set it as 2, so that we don't accidentally read outside of addressable memory.
                2
            }
            InstructionSet::A32 => 4,
            InstructionSet::A64 => 4,
            InstructionSet::RV32 => 4,
            InstructionSet::RV32C => 2,
            InstructionSet::Xtensa => 2,
        }
    }
    /// Get the maximum instruction size in bytes. All supported architectures have a maximum instruction size of 4 bytes.
    pub fn get_maximum_instruction_size(&self) -> u8 {
        // TODO: Xtensa may have wide instructions
        4
    }

    /// Returns whether a CPU with the `self` instruction set is compatible with a program compiled for `instr_set`.
    pub fn is_compatible(&self, instr_set: InstructionSet) -> bool {
        if *self == instr_set {
            return true;
        }

        matches!(
            (self, instr_set),
            (InstructionSet::RV32C, InstructionSet::RV32)
        )
    }
}

/// This describes a chip family with all its variants.
///
/// This struct is usually read from a target description
/// file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChipFamily {
    /// This is the name of the chip family in base form.
    /// E.g. `nRF52832`.
    pub name: String,
    /// The JEP106 code of the manufacturer.
    pub manufacturer: Option<JEP106Code>,
    /// The `target-gen` process will set this to `true`.
    /// Please change this to `false` if this file is modified from the generated, or is a manually created target description.
    #[serde(default)]
    pub generated_from_pack: bool,
    /// The latest release of the pack file from which this was generated.
    /// Values:
    /// - `Some("1.3.0")` if the latest pack file release was for example "1.3.0".
    /// - `None` if this was not generated from a pack file, or has been modified since it was generated.
    #[serde(default)]
    pub pack_file_release: Option<String>,
    /// This vector holds all the variants of the family.
    pub variants: Vec<Chip>,
    /// This vector holds all available algorithms.
    pub flash_algorithms: Vec<RawFlashAlgorithm>,
    #[serde(skip, default = "default_source")]
    /// Source of the target description, used for diagnostics
    pub source: TargetDescriptionSource,
}

fn default_source() -> TargetDescriptionSource {
    TargetDescriptionSource::External
}

impl ChipFamily {
    /// Loads a `ChipFamily` from a file or directory.
    pub fn load(path: impl AsRef<Path>) -> Result<Self, LoadError> {
        let path = path.as_ref();
        let value = if path.is_dir() {
            // We found a compound target definition, so we fetch all the target yamls,
            // merge them in order and load them finally.
            let mut config_parts = HashMap::new();
            let mut flash_algorithms = HashMap::new();
            for entry in
                std::fs::read_dir(path).map_err(|_| LoadError::TargetDirRead(path.into()))?
            {
                let entry = entry.map_err(LoadError::DirEntryRead)?;
                let path = entry.path();
                // We only grab yaml files. Directories are ignored.
                if let Some(extension) = path.extension() {
                    if extension.eq_ignore_ascii_case("yaml") {
                        let path_clone = path.clone();
                        let string = read_to_string(&path)
                            .map_err(|_| LoadError::TargetDefinitionFileRead(path_clone))?;
                        let yaml: Result<Value, _> = serde_yaml::from_str(&string);
                        config_parts.insert(
                            path.file_stem()
                                .map(|s: &std::ffi::OsStr| s.to_string_lossy())
                                .unwrap_or_else(|| "<unknown>".into())
                                .to_string(),
                            yaml.map_err(|parent| LoadError::TargetDefinitionFileParse {
                                path: path.clone(),
                                parent,
                            })?,
                        );
                    }

                    if extension.eq_ignore_ascii_case("elf") {
                        let path_clone = path.clone();
                        let string = read(&path)
                            .map_err(|_| LoadError::TargetDefinitionFileRead(path_clone))?;
                        flash_algorithms.insert(
                            path.file_stem()
                                .map(|s: &std::ffi::OsStr| s.to_string_lossy())
                                .unwrap_or_else(|| "<unknown>".into())
                                .to_string(),
                            string,
                        );
                    }
                }
            }
            let mut to_merge = vec![];
            if let Some(yaml) = config_parts.remove("generated") {
                to_merge.push(yaml)
            }
            for (_, yaml) in config_parts {
                to_merge.push(yaml)
            }

            let mut family: Value = Self::merge(to_merge)?;

            for (name, bytes) in flash_algorithms {
                let algorithm = family["flash_algorithms"]
                    .as_sequence_mut()
                    .unwrap()
                    .iter_mut()
                    .find(|a| a.get("name").unwrap().as_str().unwrap() == name);
                let Some(algorithm) = algorithm else {
                    return Err(LoadError::ExcessFlashAlgorithm(name));
                };
                algorithm["instructions"] =
                    Value::String(base64::engine::general_purpose::STANDARD.encode(bytes));
            }

            family
        } else if let Some(extension) = path.extension() {
            if extension.eq_ignore_ascii_case("yaml") {
                // We found a bare target yaml (old style) so we load it.
                let string = read_to_string(path)
                    .map_err(|_| LoadError::TargetDefinitionFileRead(path.into()))?;

                serde_yaml::from_str(&string).map_err(|parent| {
                    LoadError::TargetDefinitionFileParse {
                        path: path.into(),
                        parent,
                    }
                })?
            } else {
                return Err(LoadError::NotAYaml(path.into()));
            }
        } else {
            return Err(LoadError::UnrecongnizedFile(path.into()));
        };

        serde_yaml::from_value(value).map_err(LoadError::UnexpectedYamlFormat)
    }

    /// Merges all given yaml payloads.
    fn merge(files: Vec<Value>) -> Result<Value, LoadError> {
        let mut value = Value::Null;
        for new in files {
            if let Value::Mapping(new) = &new {
                if let Value::Mapping(old) = &mut value {
                    Self::merge_object(old, new);
                    continue;
                }
            }
            value = new;
        }
        Ok(value)
    }

    /// Merges two yaml mappings.
    fn merge_object(value: &mut Mapping, insert: &Mapping) {
        for (key, new) in insert.into_iter() {
            let old = value.entry(key.clone()).or_insert(Value::Null);

            if let Value::Mapping(old) = old {
                if let Value::Mapping(new) = new {
                    Self::merge_object(old, new);
                    continue;
                }
            }
            if &Value::Null == new {
                value.remove(key);
                continue;
            }
            *old = new.clone();
        }
    }

    /// Validates the [`ChipFamily`] such that probe-rs can make assumptions about the correctness without validating thereafter.
    ///
    /// This method should be called right after the [`ChipFamily`] is created!
    pub fn validate(&self) -> Result<(), String> {
        // We check each variant if it is valid.
        // If one is not valid, we abort with an appropriate error message.
        for variant in &self.variants {
            // Make sure the algorithms used on the variant actually exist on the family (this is basically a check for typos).
            for algorithm_name in variant.flash_algorithms.iter() {
                if !self
                    .flash_algorithms
                    .iter()
                    .any(|algorithm| &algorithm.name == algorithm_name)
                {
                    return Err(format!(
                        "unknown flash algorithm `{}` for variant `{}`",
                        algorithm_name, variant.name
                    ));
                }
            }

            // Check that there is at least one core.
            if let Some(core) = variant.cores.first() {
                // Make sure that the core types (architectures) are not mixed.
                let architecture = core.core_type.architecture();
                if variant
                    .cores
                    .iter()
                    .any(|core| core.core_type.architecture() != architecture)
                {
                    return Err(format!(
                        "definition for variant `{}` contains mixed core architectures",
                        variant.name
                    ));
                }
            } else {
                return Err(format!(
                    "definition for variant `{}` does not contain any cores",
                    variant.name
                ));
            }

            // Core specific validation logic based on type
            for core in variant.cores.iter() {
                // The core access options must match the core type specified
                match &core.core_access_options {
                    CoreAccessOptions::Arm(options) => {
                        if !matches!(
                            core.core_type,
                            CoreType::Armv6m
                                | CoreType::Armv7a
                                | CoreType::Armv7em
                                | CoreType::Armv7m
                                | CoreType::Armv8a
                                | CoreType::Armv8m
                        ) {
                            return Err(format!(
                                "Arm options don't match core type {:?} on core {}",
                                core.core_type, core.name
                            ));
                        }

                        if matches!(core.core_type, CoreType::Armv7a | CoreType::Armv8a)
                            && options.debug_base.is_none()
                        {
                            return Err(format!("Core {} requires setting debug_base", core.name));
                        }

                        if core.core_type == CoreType::Armv8a && options.cti_base.is_none() {
                            return Err(format!("Core {} requires setting cti_base", core.name));
                        }
                    }
                    CoreAccessOptions::Riscv(_) => {
                        if core.core_type != CoreType::Riscv {
                            return Err(format!(
                                "Riscv options don't match core type {:?} on core {}",
                                core.core_type, core.name
                            ));
                        }
                    }
                    CoreAccessOptions::Xtensa(_) => {
                        if core.core_type != CoreType::Xtensa {
                            return Err(format!(
                                "Xtensa options don't match core type {:?} on core {}",
                                core.core_type, core.name
                            ));
                        }
                    }
                }
            }

            let core_names: Vec<_> = variant.cores.iter().map(|core| &core.name).collect();

            for memory in &variant.memory_map {
                // Ensure that the memory is assigned to a core, and that all the cores exist

                for core in memory.cores() {
                    if !core_names.contains(&core) {
                        return Err(format!(
                            "Variant {}, memory region {:?} is assigned to a non-existent core {}",
                            variant.name, memory, core
                        ));
                    }
                }

                assert!(
                    !memory.cores().is_empty(),
                    "Variant {}, memory region {:?} is not assigned to a core",
                    variant.name,
                    memory
                );
            }
        }

        Ok(())
    }

    /// Get the different [Chip]s which are part of this
    /// family.
    pub fn variants(&self) -> &[Chip] {
        &self.variants
    }

    /// Get all flash algorithms for this family of chips.
    pub fn algorithms(&self) -> &[RawFlashAlgorithm] {
        &self.flash_algorithms
    }

    /// Try to find a [RawFlashAlgorithm] with a given name.
    pub fn get_algorithm(&self, name: impl AsRef<str>) -> Option<&RawFlashAlgorithm> {
        let name = name.as_ref();
        self.flash_algorithms.iter().find(|elem| elem.name == name)
    }
}

/// An error while loading a chip family ocurred.
#[derive(Debug, thiserror::Error)]
pub enum LoadError {
    #[error("flash algorithm ELF ({0}) found without an entry in the yaml")]
    ExcessFlashAlgorithm(String),
    #[error("target dir at {0:?} could not be read")]
    TargetDirRead(PathBuf),
    #[error("dir entry could not be read")]
    DirEntryRead(#[source] std::io::Error),
    #[error("target definition file at {0:?} could not be read")]
    TargetDefinitionFileRead(PathBuf),
    #[error("target definition file at {path:?} could not be parsed")]
    TargetDefinitionFileParse {
        path: PathBuf,
        #[source]
        parent: serde_yaml::Error,
    },
    #[error("target file at {0:?} is not a yaml")]
    NotAYaml(PathBuf),
    #[error("target file at {0:?} could not be read")]
    TargetFileRead(PathBuf),
    #[error("unrecognized target file at {0:?}")]
    UnrecongnizedFile(PathBuf),
    #[error("The yaml format did not match the expected one")]
    UnexpectedYamlFormat(#[source] serde_yaml::Error),
}

#[cfg(test)]
mod tests {
    use crate::ChipFamily;

    #[test]
    fn basic_merge() {
        let values = vec![
            serde_yaml::from_str(r#"{ a: 65, b: { b: 66 }, c: { c: 67 } }"#).unwrap(),
            serde_yaml::from_str(r#"{ a: null, b: { a: 65 }, c: 67 }"#).unwrap(),
        ];

        let merged = ChipFamily::merge(values).unwrap();
        insta::assert_yaml_snapshot!(merged);
    }
}
