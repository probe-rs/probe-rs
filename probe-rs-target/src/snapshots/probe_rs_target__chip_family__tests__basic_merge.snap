---
source: probe-rs-target/src/chip_family.rs
assertion_line: 428
expression: merged
---
c: 67
b:
  b: 66
  a: 65

